Introduce integ tests for high disk watermark

DaveCTurner opened the pull request:
An important goal of the disk threshold decider is to ensure that nodes
use less disk space than the high watermark, and to take action if a
node ever exceeds this watermark. Today we do not have any
integration-style tests of this high-level behaviour. This commit
introduces a small test harness that can adjust the apparent size of the
disk and verify that the disk threshold decider moves shards around in
response.

DaveCTurner committed:
Introduce integ tests for high disk watermark

An important goal of the disk threshold decider is to ensure that nodes
use less disk space than the high watermark, and to take action if a
node ever exceeds this watermark. Today we do not have any
integration-style tests of this high-level behaviour. This commit
introduces a small test harness that can adjust the apparent size of the
disk and verify that the disk threshold decider moves shards around in
response.

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Allocation)

DaveCTurner committed:
Defeat the is-spinning-disk check

DaveCTurner committed:
More hacker

DaveCTurner commented:
Ugh this works fine on my laptop now but Linux is a different matter entirely it seems. Needs more work so I've marked it as a draft.

DaveCTurner committed:
Mock filesystem at suite level, don't replace it every test

DaveCTurner committed:
Merge branch 'master' into 2020-07-30-high-watermark-integ-tests-WIP

DaveCTurner committed:
Wrong IOUtils

ywelsch commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderIT.java:
```diff
@@ -0,0 +1,339 @@
+                indexRequestBuilders[i] = client().prepareIndex("test").setSource("field", randomAlphaOfLength(10));
+            }
+            indexRandom(true, indexRequestBuilders);
+            forceMerge();
```
why the force-merge?

ywelsch replied:
seems not to work otherwise. Not really relevant whether or not we  force-merge, so I left as is

ywelsch commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderIT.java:
```diff
@@ -0,0 +1,339 @@
+        ((InternalClusterInfoService) internalCluster().getMasterNodeInstance(ClusterInfoService.class)).refresh();
+        // if the nodes were all under the low watermark already (but unbalanced) then a change in the disk usage doesn't trigger a reroute
+        // even though it's now possible to achieve better balance, so we have to do an explicit reroute. TODO fix this?
+        assertAcked(client().admin().cluster().prepareReroute());
```
can we somehow only conditionally do this reroute here, i.e.  if all nodes under low watermark?

ywelsch commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderIT.java:
```diff
@@ -0,0 +1,339 @@
+                    // probably removed
+                    return 0L;
+                }
+            } else if (path.getFileName().toString().equals("_state") || path.getFileName().toString().equals("translog")) {
```
Instead of excluding _state and  translog path, should we select the index folder explicitly? Or  assert that  there are  no other paths?

ywelsch replied:
we recursively step into these directories from the root path. including folder is therefore  much trickier

ywelsch reviewed (commented):
Minor comments

ywelsch committed:
Merge remote-tracking branch 'elastic/master' into high-disk-tests

ywelsch committed:
Fix issue with overriding filesystem

When tests were run in a loop, e.g. using

./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.cluster.routing.allocation.decider.DiskThresholdDeciderIT.testHighWatermarkNotExceeded" -Dtests.seed=CBF5914690DC5CAC -Dtests.security.manager=true -Dtests.jvms=4 -Dtests.locale=de-CH -Dtests.timezone=Asia/Chungking -Dtests.iters=2 -Dtests.failfast=true

they would fail due to the way filesystem are overridden, interfering with the filesystem setup done by ESTestCase

ywelsch committed:
conditionally reroute

ywelsch committed:
Merge remote-tracking branch 'elastic/master' into high-disk-tests

ywelsch commented:
I've updated the PR and addressed my remaining comments. Can you have a look @tlrx? I'll need this for  #61906

tlrx reviewed (approved):
LGTM

ywelsch commented:
David says thanks!
